### PR TITLE
Uniform distribution implemented

### PR DESCRIPTION
**Pull request by kirillbilchenko**
Uniform distribution of load between available hosts

### CODE REVIEW

**Comment by RTodorov** [comment]
@kirillbilchenko code coverage for that project is about 90%, if you intend to merge this upstream, they'll probably demand some tests

File: config.go [hunk 1]
```diff
@@ -97,3 +110,29 @@ func choose(hosts Hosts) (*Host, int) {
...
 	return nil, -1
 }
+
+func findFirstNodeByQueue(hosts Hosts, mutex *sync.RWMutex) Host {
+	var capacities []capacity
+	mutex.Lock()
+	defer mutex.Unlock()
+	for _, y := range hosts {
+		rsp, _ := http.Get(y.StatusEndPoint())
+		if rsp != nil {
```
Comment (antonu17) [inline config.go]: Here rsp might not be `nil` even if http.Get returned an error. Next ioutil.ReadAll(rsp.Body) will fail with some other error. but function will still continue to execute. Please add error handling.

**Comment by kirillbilchenko** [comment]
@RTodorov I update the tests, can you please take a look again

File: config.go [hunk 2]
```diff
@@ -97,3 +113,49 @@ func choose(hosts Hosts) (*Host, int) {
...
+	}
+	var capacities []capacity
+	mutex.Lock()
+	defer mutex.Unlock()
+	var netClient = &http.Client{
+		Timeout: time.Second * 10,
+	}
+	for y := range *hosts {
+		rsp, err := netClient.Get((*hosts)[y].StatusEndPoint())
+		if err == nil {
```
Comment (aleksandrzhiliaev) [inline config.go]: do we need to do anything if `err!=nil`?

File: config.go [hunk 3]
```diff
@@ -97,3 +113,49 @@ func choose(hosts Hosts) (*Host, int) {
...
+	for v := range *hosts {
+		hostMap[fmt.Sprintf("%s%d%s", (*hosts)[v].Name, (*hosts)[v].Port, (*hosts)[v].Region)] = &(*hosts)[v]
+	}
+	var capacities []capacity
+	mutex.Lock()
+	defer mutex.Unlock()
+	var netClient = &http.Client{
+		Timeout: time.Second * 10,
+	}
+	for y := range *hosts {
```
Comment (aleksandrzhiliaev) [inline config.go]: maybe `h`?
```suggestion
	for h := range *hosts {
```

File: config.go [hunk 4]
```diff
@@ -97,3 +113,49 @@ func choose(hosts Hosts) (*Host, int) {
...
+		rsp, err := netClient.Get((*hosts)[y].StatusEndPoint())
+		if err == nil {
+			responseMap := map[string]int{}
+			body, err := ioutil.ReadAll(rsp.Body)
+			if err == nil {
+				_ = json.Unmarshal(body, &responseMap)
```
Comment (aleksandrzhiliaev) [inline config.go]: you don't handle `json.Unmarshal` error, if `body` is invalid it panics

File: config.go [hunk 5]
```diff
@@ -97,3 +113,49 @@ func choose(hosts Hosts) (*Host, int) {
...
+				capacities = append(capacities, capacity{&(*hosts)[y], responseMap["queued"], responseMap["pending"], responseMap["used"], responseMap["total"]})
```
Comment (aleksandrzhiliaev) [inline config.go]: what if there no `queued`, `pending`, `used` or `total` field in json?
You should validate it first, or unmarshal to your `capacity` struct directly.

**Pull request merged**